File: Kalorie/Kalorie/Core/UseCases/FetchFoodsConsumedInRangeUseCase.swift
//
//  FetchFoodsConsumedInRangeUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 19.09.2026.
//

import Foundation

protocol FetchFoodsConsumedInRangeUseCaseProtocol {
    func callAsFunction(from: Date, to: Date) async throws -> [FoodConsumedDomain]
}

struct FetchFoodsConsumedInRangeUseCase: FetchFoodsConsumedInRangeUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(from: Date, to: Date) async throws -> [FoodConsumedDomain] {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let calendar = Calendar.current
        let start = calendar.startOfDay(for: from)
        let endDay = calendar.startOfDay(for: to)
        let end = calendar.date(byAdding: .day, value: 1, to: endDay) ?? endDay.addingTimeInterval(Constants.Time.secondsPerDay)
        let dtos: [FoodConsumedDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.foodConsumed(userId: userId),
            where: "date",
            isGreaterThanOrEqualTo: start.timeIntervalSince1970,
            isLessThan: end.timeIntervalSince1970
        )
        return dtos.map { $0.asDomain() }
    }
}

#if DEBUG
struct FetchFoodsConsumedInRangeUseCaseFake: FetchFoodsConsumedInRangeUseCaseProtocol {

    // MARK: - Properties

    var stubbedFoods: [FoodConsumedDomain] = []
    var stubbedError: Error?

    // MARK: - Functions

    func callAsFunction(from: Date, to: Date) async throws -> [FoodConsumedDomain] {
        if let stubbedError { throw stubbedError }
        return stubbedFoods
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchFoodsConsumedInRangeUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodConsumedDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodConsumedDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants
import java.time.Instant
import java.time.ZoneId

interface FetchFoodsConsumedInRangeUseCaseProtocol {
    suspend operator fun invoke(from: Instant, to: Instant): List<FoodConsumedDomain>
}

class FetchFoodsConsumedInRangeUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
    private val zone: ZoneId = ZoneId.systemDefault(),
) : FetchFoodsConsumedInRangeUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(from: Instant, to: Instant): List<FoodConsumedDomain> {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val start = from.atZone(zone).toLocalDate().atStartOfDay(zone)
        val end = to.atZone(zone).toLocalDate().plusDays(1).atStartOfDay(zone)
        val dtos: List<FoodConsumedDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.foodConsumed(userId),
            field = "date",
            isGreaterThanOrEqualTo = start.toEpochSecond().toDouble(),
            isLessThan = end.toEpochSecond().toDouble(),
        )
        return dtos.map { it.asDomain() }
    }
}
